src/main/kotlin/com/yapp/itemfinder/domain/item/ItemRepository.kt

```diff
@@ -1,5 +1,71 @@
 package com.yapp.itemfinder.domain.item
 
+import com.querydsl.core.types.dsl.BooleanExpression
+import com.querydsl.jpa.impl.JPAQueryFactory
+import com.yapp.itemfinder.domain.item.QItemEntity.itemEntity
+import com.yapp.itemfinder.domain.item.dto.ItemSearchOption
+import com.yapp.itemfinder.domain.tag.QItemTagEntity.itemTagEntity
+import com.yapp.itemfinder.domain.tag.QTagEntity.tagEntity
+import com.yapp.itemfinder.support.PaginationHelper
+import org.springframework.data.domain.Page
+import org.springframework.data.domain.Pageable
 import org.springframework.data.jpa.repository.JpaRepository
 
-interface ItemRepository : JpaRepository<ItemEntity, Long>
+interface ItemRepository : JpaRepository<ItemEntity, Long>, ItemRepositorySupport
+interface ItemRepositorySupport {
+    fun search(searchOption: ItemSearchOption, pageable: Pageable, targetContainerIds: List<Long>): Page<ItemEntity>
+}
+
+class ItemRepositorySupportImpl(
+    private val queryFactory: JPAQueryFactory,
+    private val paginationHelper: PaginationHelper
+) : ItemRepositorySupport {
+    override fun search(searchOption: ItemSearchOption, pageable: Pageable, targetContainerIds: List<Long>): Page<ItemEntity> {
+        val query = queryFactory.select(itemEntity)
+            .from(itemEntity)
+            .leftJoin(itemEntity.tags, itemTagEntity)
+            .leftJoin(itemTagEntity.tag, tagEntity)
+            .where(
+                containerIdIsIn(targetContainerIds),
+                itemTypeIsIn(searchOption.itemTypes),
+                tagNameIsIn(searchOption.tagNames),
+                itemNameContains(searchOption.itemName)
+            )
+            .groupBy(itemEntity.id)
+            .having(tagEntity.count().goe(searchOption.tagNames.size))
+
+        return paginationHelper.getPage(pageable, query, ItemEntity::class.java)
+    }
+
+    private fun containerIdIsIn(targetContainerIds: List<Long>): BooleanExpression? {
+        return if (targetContainerIds.isNotEmpty()) {
+            itemEntity.container.id.`in`(targetContainerIds)
+        } else {
+            null
+        }
+    }
```

Comment:
@osandra [여기](https://github.com/YAPP-Github/21st-Android-Team-2-BE/pull/21#discussion_r1089904413)에서 말씀드린 이유로 다음과 같이 수정하면 좋겠습니다.

```suggestion
    private fun containerIdIsIn(targetContainerIds: List<Long>): BooleanExpression {
         return itemEntity.container.id.`in`(targetContainerIds)
    }
```

Reply:
감사합니다!! 해당 부분 수정했습니다 :)) [8ce3e00](https://github.com/YAPP-Github/21st-Android-Team-2-BE/pull/21/commits/8ce3e0052a8d776354b5df339ef95df0796a74b3)